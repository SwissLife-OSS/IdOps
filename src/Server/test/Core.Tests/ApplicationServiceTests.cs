using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using FluentAssertions;
using IdOps.Model;
using Moq;
using Snapshooter.Xunit;
using Xunit;

namespace IdOps.Core.Tests
{
    public class ApplicationServiceTests : ServiceTestsBase
    {
        [Fact]
        public async Task Save_WithValidRequest_IsSaved()
        {
            // Arrange
            Guid envId = Guid.Parse("e1579ce6-b8c3-4479-acf1-f35bbeee8b5b");

            IEnumerable<Model.Environment> environments =
                new List<Model.Environment>
                {
                    new Model.Environment { Id = envId, Name = "TestEnv", Order = 0 }
                };

            Mock<IResourceManager> resourceManagerMock = CreateResourceManager_ApplicationMock();

            Mock<IEnvironmentService> environmentServiceMock = CreateEnvironmentServiceMock(
                environments,
                (environmentList) =>
                {
                    environments = environmentList;
                });

            string tenant = "Contoso";

            ICollection<string> allowedGrantTypes = new List<string>();
            allowedGrantTypes.Add("asdf");

            ICollection<string> redirectUris = new List<string>();
            redirectUris.Add("https://asdf.ch");

            ICollection<Guid> env = new List<Guid>();
            env.Add(envId);

            var service = new ApplicationService(
                applicationStore: default,
                clientStore: default,
                userContextAccessor: default,
                resourceManager: resourceManagerMock.Object,
                clientService: CreateClientServiceMock().Object,
                clientTemplateService: CreateClientTemplateServiceMock().Object,
                environmentService: environmentServiceMock.Object);

            var request = new CreateApplicationRequest(
                Name: "MyApplication",
                Tenant: tenant,
                TemplateId: Guid.Parse("5bc2dd66f14245fca7a7e3b725f90147"),
                AllowedGrantTypes: allowedGrantTypes.ToArray(),
                ApiScopes: CreateApiScopes().ToArray(),
                IdentityScopes: CreateIdentityScopes().ToArray(),
                Environments: env.ToArray(),
                RedirectUris: redirectUris.ToArray());

            // Act
            ApplicationWithClients application = await service.CreateAsync(request, default);

            // Assert
            environmentServiceMock.Verify(m => m.GetAllAsync(It.IsAny<CancellationToken>()),
                Times.Once);

            application.Application.Tenant.Should().Be(tenant);
            application.Application.AllowedGrantTypes.Should().BeEquivalentTo(allowedGrantTypes);
            application.Application.RedirectUris.Should().BeEquivalentTo(redirectUris);


            ICollection<CreatedClientInfo> clients = (List<CreatedClientInfo>)application.Clients;
            CreatedClientInfo client = clients.FirstOrDefault();
            application.Application.ClientIds.Should().Contain(client.Id);

            application.MatchSnapshot(x => x.IgnoreField("**.Id").IgnoreField("**.ClientIds"));
        }
    }
}
