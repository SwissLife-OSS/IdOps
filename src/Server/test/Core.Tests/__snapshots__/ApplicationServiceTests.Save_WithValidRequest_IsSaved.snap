{
  "Application": {
    "Id": "aa69f6a7-e5bf-48c7-a6e2-3371d2503f53",
    "Tenant": "Contoso",
    "Name": "MyApplication",
    "Title": "MyApplication",
    "ApiScopes": [
      "202badbf-7922-4832-8651-ac7984d44774",
      "5772ee6f-216c-4298-855c-42d64b946a40"
    ],
    "IdentityScopes": [
      "5135561c-6827-40c8-9f0f-3ffe8d1d9fbe"
    ],
    "RedirectUris": [
      "https://asdf.ch"
    ],
    "PostLogoutRedirectUris": [],
    "ClientIds": [
      "d72c1f59-3244-4bb4-add7-5a5da9b272b4"
    ],
    "Version": null,
    "AllowedGrantTypes": [
      "asdf"
    ],
    "TemplateId": "5bc2dd66-f142-45fc-a7a7-e3b725f90147"
  },
  "Clients": [
    {
      "Id": "d72c1f59-3244-4bb4-add7-5a5da9b272b4",
      "ClientId": null,
      "Name": null,
      "SecretValue": "fdasdfasdf"
    }
  ]
}
